cli/lsp/performance.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::serde::Deserialize;
+use deno_core::serde::Serialize;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::Duration;
+use std::time::Instant;
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PerformanceAverage {
+  pub name: String,
+  pub count: u32,
+  pub average_duration: u32,
+}
+
+/// A structure which serves as a start of a measurement span.
+#[derive(Debug)]
+pub struct PerformanceMark {
+  name: String,
+  count: u32,
+  start: Instant,
+}
+
+/// A structure which holds the information about the measured span.
+#[derive(Debug, Clone)]
+pub struct PerformanceMeasure {
+  pub name: String,
+  pub count: u32,
+  pub duration: Duration,
+}
+
+impl From<PerformanceMark> for PerformanceMeasure {
+  fn from(value: PerformanceMark) -> Self {
+    Self {
+      name: value.name,
+      count: value.count,
+      duration: value.start.elapsed(),
+    }
+  }
+}
+
+/// A simple structure for marking a start of something to measure the duration
+/// of and measuring that duration.  Each measurement is identified by a string
+/// name and a counter is incremented each time a new measurement is marked.
+///
+/// The structure will limit the size of measurements to the most recent 1000,
+/// and will roll off when that limit is reached.
+#[derive(Debug, Clone)]
+pub struct Performance {
+  counts: Arc<Mutex<HashMap<String, u32>>>,
+  measures: Arc<Mutex<VecDeque<PerformanceMeasure>>>,
+  size: usize,
+}
+
+impl Default for Performance {
+  fn default() -> Self {
+    Self {
+      counts: Default::default(),
+      measures: Default::default(),
+      size: 1_000,
+    }
+  }
+}
+
+impl Performance {
+  /// Return the count and average duration of a measurement identified by name.
+  #[allow(unused)]
+  pub fn average(&self, name: &str) -> (usize, Duration) {
+    let mut items = Vec::new();
+    for measure in self.measures.lock().unwrap().iter() {
+      if measure.name == name {
+        items.push(measure.duration);
+      }
+    }
+    let len = items.len();
+    let average = if len > 0 {
+      items.into_iter().sum::<Duration>() / len as u32
+    } else {
+      Duration::default()
+    };
+    (len, average)
+  }
+
+  /// Return an iterator which provides the names, count, and average duration
+  /// of each measurement.
+  pub fn averages(&self) -> impl Iterator<Item = (String, usize, Duration)> {
+    let mut averages: HashMap<String, Vec<Duration>> = HashMap::new();
+    for measure in self.measures.lock().unwrap().iter() {
+      averages
+        .entry(measure.name.clone())
+        .or_default()
+        .push(measure.duration);
+    }
+    averages.into_iter().map(|(k, d)| {
+      let a = d.clone().into_iter().sum::<Duration>() / d.len() as u32;
```

Comment:
I'm a bit worried this is integer division which gets rounded... Does  `d.clone().into_iter().sum::<Duration>() ` have the type `Duration`? What does it mean to divide a Duration by a u32 ?

Reply:
`Duration` has a `Div<u32>` trait.  It is the appropriate of "averaging" as `Duration`.